src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -105,14 +106,22 @@ async Task<Solution> GetInitialUncleanedSolutionAsync(Solution originalSolution)
                     // expensive as we'd fork, produce semantics, fork, produce semantics, etc. etc.). Instead, by
                     // adding all the changed documents to one solution, and then cleaning *those* we only perform
                     // cleanup semantics on one forked solution.
-                    await foreach (var (docId, (newRoot, newText)) in stream)
+
+                    using var _1 = ArrayBuilder<(DocumentId documentId, SyntaxNode newRoot, PreservationMode mode)>.GetInstance(out var syntaxRoots);
+                    using var _2 = ArrayBuilder<(DocumentId documentId, SourceText newText, PreservationMode mode)>.GetInstance(out var newTexts);
+
+                    await foreach (var (documentId, (newRoot, newText)) in stream)
                     {
-                        currentSolution = newRoot != null
-                            ? currentSolution.WithDocumentSyntaxRoot(docId, newRoot)
-                            : currentSolution.WithDocumentText(docId, newText!);
+                        if (newRoot != null)
+                            syntaxRoots.Add((documentId, newRoot, PreservationMode.PreserveValue));
+                        else
+                            newTexts.Add((documentId, newText!, PreservationMode.PreserveValue));
                     }
 
-                    return currentSolution;
+                    return args.originalSolution
+                        .WithDocumentSyntaxRoots(syntaxRoots.ToImmutableAndClear())
+                        .WithDocumentTexts(newTexts.ToImmutableAndClear());
```

Comment:
>WithDocumentTexts [](http://example.com/codeflow?start=25&length=17)

I feel like there are a bunch of places that could benefit from this. Have you done an audit for other possible changes. OK with that being separated from this PR.

Reply:
i audited the places i just added :)  i will also look for other usages of WithDocumentSyntaxRoot and see if any cleanly can move over.